### ISSUE TITLE
terra::as.polygons() does not behave as raster::rasterToPolygons()

### ISSUE THREAD

**Issue by hrvg** [rspatial/terra#40]
I am using `terra` version 0.6.9 (alpha-release) and `terra::as.polygons()` does not behave as `raster::rasterToPolygons()` which is what I expected from the documentation.

I was tiling a DEM of California when I run into this issue by noticing that some of the tile had larger size than expected. Here is the source file: [CA_DEM_Lmax-resampled.zip](https://github.com/rspatial/terra/files/4761179/CA_DEM_Lmax-resampled.zip)

You can notice that the order of the polygons in `raster::rasterToPolygons()` is conform with what I expected (starting in top left corner). It's not the case with `terra::as.polygons()`. The number of extracted polygons is also different between the two packages. When using polygons from `terra`, some geometries have a larger extent (red and blue case below) resulting in larger tiled DEM.

```
par(mfrow = c(1, 2))

library("raster")
resampled_raster <- raster:::raster("CA_DEM_Lmax-resampled.tif")
pol_raster <- raster::rasterToPolygons(resampled_raster)
plot(resampled_raster, alpha = 0.4)
plot(pol_raster, add = T)
plot(pol_raster[1, ], col = "black", add = T)
plot(pol_raster[2, ], col = "gray", add = T)
plot(pol_raster[3, ], col = "lightgray", add = T)
plot(pol_raster[24, ], col = "blue", add = T)
plot(pol_raster[31, ], col = "red", add = T)

plot(maxL_polygons)

library("terra")
resampled_raster <- terra:::rast("CA_DEM_Lmax-resampled.tif")
pol_terra <- terra::as.polygons(resampled_raster)
plot(resampled_raster, alpha = 0.4)
plot(pol_terra, add = T)
plot(pol_terra[1, ], col = "black", add = T)
plot(pol_terra[2, ], col = "gray", add = T)
plot(pol_terra[3, ], col = "lightgray", add = T)
plot(pol_terra[24, ], col = "blue", add = T)
plot(pol_terra[31, ], col = "red", add = T)

length(pol_raster) == length(pol_terra)
```

**Comment by rhijmans** [comment]
Thank you, this is very helpful. Here is how you would get the same result (by using `dissolve=FALSE`)

```
par(mfrow = c(1, 2))
plotstuff <- function(pol) {
	plot(pol, add = T)
	plot(pol[1, ], col = "black", add = T)
	plot(pol[2, ], col = "gray", add = T)
	plot(pol[3, ], col = "lightgray", add = T)
	plot(pol[24, ], col = "blue", add = T)
	plot(pol[31, ], col = "red", add = T)
}

library("raster")
resampled_raster <- raster:::raster("CA_DEM_Lmax-resampled.tif")
pol_raster <- raster::rasterToPolygons(resampled_raster)
plot(resampled_raster)
plotstuff(pol_raster)

library("terra")
resampled_terra <- terra:::rast("CA_DEM_Lmax-resampled.tif")
pol_terra <- terra::as.polygons(resampled_terra, dissolve=FALSE)
plot(resampled_terra)
plotstuff(pol_terra)
```

Using `dissolve=FALSE` is considerably slower, but at this resolution that does not matter. You could also do 

```
pol_terra <- terra::as.polygons(resampled_terra, dissolve=TRUE)
pol_terra <- disaggregate(pol_terra)
```
That would work in this case, but only because the aggregated polygons are multipolygons (i.e. they have no shared boundaries, so the polygons were not merged). And the order is not as expected (not as in raster, and no guaranteed order). 

Moreover, this also reveals a bug in `terra`: cells with a value of zero are not ignored as if they were `NA`. I will leave this issue open until that has been fixed.

**Comment by hrvg** [comment]
Thanks for this great answer. 

I was looking at an older version of the documentation which did not mention the `dissolve` option. Documentation for 0.6.9 mentions the option but Google landed me on the documentation for 0.5.8.

From your answer, `terra` defaults to `dissolve=TRUE` for speed, while `raster` default to `dissolve=FALSE` which is why similar functions in the two packages produce different results. A warning in code or documentation might be useful for users porting code from `raster` to `terra`.

**Comment by rhijmans** [comment]
The bug has now been fixed. I have also added a section to the overview page in the manual (`?terra` ) that lists differences in the behavior of functions with the same name (or that are conceptually similar such as `rasterToPolygons` and `as.polygons`) between `raster` and `terra`. Thank you for that suggestion.
